### ISSUE TITLE
link problem 

### ISSUE THREAD

**Issue by ryniker** [minoca/os#76]
ln command exit code 44  ENOTSUP

~$ ls -l foo zoo
ls -l foo zoo
ls: Cannot stat 'zoo': No such file or directory.
-rw-r--r--  1 ryniker ryniker  0 Dec 04 20:42 foo
~$ ln foo zoo
ln foo zoo
~$ echo $?
echo $?
44
~$ ls -l foo zoo
ls -l foo zoo
ls: Cannot stat 'zoo': No such file or directory.
-rw-r--r--  1 ryniker ryniker  0 Dec 04 20:42 foo
~$

**Comment by drawkula** [comment]
That's mentioned in https://github.com/minoca/os/issues/24#issuecomment-260022318 .

**Comment by evangreen** [comment]
We were able to hack FAT to accommodate a lot of things, like Unix-style permissions, symlinks, and user/group IDs. In fact, we're able to do this in a way that's transparent to other OSes that might use the same file system. But we couldn't quite do hard links in a way that wouldn't spectacularly explode if you operated the drive on another OS. Since interoperability is the biggest appeal of FAT, we left this one as "not supported"

The trouble is that Unix-compatible file systems have separate concepts for Directory Entries (often called d-entries) and the actual file contents (called an i-node). Hard links work by pointing multiple directory entries at the same i-node, and increasing an internal reference count on the i-node.

FAT does not have these two concepts isolated, so the file content information is stored inside the directory entry. There is no i-node reference count that can be incremented. What this means is that creating a hard link would be easy: just point multiple directory entries at the same file contents. The problem is when you delete a file, the system would have no way of knowing that there are additional references to that file, and would free the file contents. The file system is essentially corrupt at this point.

So, long story long, to implement hard links we'll need to implement support for a new file system. We're probably leaning towards ext 2/3/4, as it's relatively simple (at least 2 is) and has pretty good performance.

**Comment by ryniker** [comment]
>In fact, we're able to do this in a way that's transparent to other OSes that might use the same file >system. But we couldn't quite do hard links

It would be good for someone to go through the Minoca documentation to at least put comments on pages like http://www.minocacorp.com/doc/1375/api/LIBC_API/element/363224/
to say the link function described is not implemented. 

I bumped into this building emacs 25.1.  it would be possible for the autoconf tools to check for link working.  I suspect either it was never considered necessary, or other checks indicated an environment where it was assumed link was present.

I solved the problem in the emacs build, but many applications use autoconf and some may have problems more difficult to resolve.  ext2 support would be a good solution.

"Transparent" is a bold claim.  If you try fsck.vfat, you will see lots of complaints.  Do you mean I can use Linux or Windows to copy the files from a Minoca filesystem to a new volume and this new volume will boot and run Minoca without trouble?  That would be very useful, because then one could expand the Minoca image to fill a volume without a need to write unused blocks using something like dd.

Even Minoca tools such as third-party/build/opkg-utils/opkg-extract-data, that you suggested and which helped me, wants to use link - and fails on Minoca volumes.

**Comment by evangreen** [comment]
You're right, that would be a good idea. Right now the lack of hard links is only documented in the body of a function somewhere.

Emacs does seem to rely on hard links doesn't it. I changed a few of those to cp -p in Makefile.in. Were there others that you tweaked as well? (You can use `make recreate-diff` in third-party/build/emacs-24.5 to regenerate the patch with any changes you made). If you've got a better set of patches I'd love to incorporate them.

I hadn't tried fsck.vfat, as I don't normally operate on Linux. Do you have the output from that you might be able to throw up to pastebin? I'd be interested to see what it's complaining about.

What I really meant by transparent is that we're able to add permissions, symlinks, and user/group owners in a way that the data generally stays with the file, but other OSes can still see and manipulate those files normally. We do this by using a "long name" always, and then encoding the Unix-ish properties into the short name.

On a raspberry pi, you might be able to get away with just copying the files to a new FAT disk. The thing stopping you from just dragging the files to a new FAT disk and booting from it on a PC is the MBR sector at the beginning of the disk and the VBR sectors at the beginning of the partition. These are needed to boot from a PC properly. 

Our msetup program is what we use to create images that have these sectors written properly. Eventually you will be able to use msetup to install directly to something like /dev/sda on Linux. The missing support on Linux is figuring out how big the disk is. If you specified that manually using -G, then I think you might be able to quickly write out an image to a larger disk using msetup, without having to dd all those empty sectors. But again, you can also try the manual cp method too.

On an unrelated note, I pushed some changes to third-party, which should hopefully fix your GCC-6 `make tools` issues with GCC, ncurses, and pkg-config. I also added an "install-minimal" make target to third-party, which compiles and extracts the minimum set of packages needed for opkg and an SSH server. If you have a moment, pull in the latest from third-party and os and give those a try.

**Comment by ryniker** [comment]
Here is output from fsck.vfat on the two file systems in Minoca-rpi-1944/rpi.img.  The first is just the
Raspberry Pi boot stuff, and fsck has no issues, which is expected.  The second is the Minoca OS
file system, which you will find of greater interest.

[ryniker@puget ~]$ fsck.vfat -nv /dev/loop0p1
fsck.fat 3.0.28 (2015-05-16)
Checking we can access the last sector of the filesystem
Boot sector contents:
System ID "MSDOS5.0"
Media byte 0xf8 (hard disk)
       512 bytes per logical sector
      4096 bytes per cluster
        32 reserved sectors
First FAT starts at byte 16384 (sector 32)
         2 FATs, 16 bit entries
     12288 bytes per FAT (= 24 sectors)
Root directory starts at byte 40960 (sector 80)
       512 root directory entries
Data area starts at byte 57344 (sector 112)
      5106 data clusters (20914176 bytes)
63 sectors/track, 128 heads
         0 hidden sectors
     40960 sectors total
Root contains directory ".". Dropping it.
Checking for unused clusters.
Leaving filesystem unchanged.
/dev/loop0p1: 11 files, 1179/5106 clusters

[ryniker@puget ~]$ fsck.vfat -nv /dev/loop0p2
fsck.fat 3.0.28 (2015-05-16)
Checking we can access the last sector of the filesystem
There is no backup boot sector.
  Auto-creating backup boot block.
Created backup of boot sector in sector 6
Boot sector contents:
System ID "MSDOS5.0"
Media byte 0xf8 (hard disk)
       512 bytes per logical sector
      4096 bytes per cluster
        32 reserved sectors
First FAT starts at byte 16384 (sector 32)
         2 FATs, 32 bit entries
    503808 bytes per FAT (= 984 sectors)
Root directory start at cluster 2 (arbitrary size)
Data area starts at byte 1024000 (sector 2000)
    125446 data clusters (513826816 bytes)
63 sectors/track, 128 heads
         0 hidden sectors
   1005568 sectors total
Root contains directory ".". Dropping it.
/pagefile.sys
  File size is 51380224 bytes, cluster chain length is > 51380224 bytes.
  Truncating file to 51380224 bytes.
/apps/bin/netcon
  Bad short file name (\213\200\200\200\200\200\200.\200\203).
  Auto-renaming it.
  Renamed to FSCK0000.000
/apps/etc/ssl/misc/CA.pl
  Bad short file name (\200\200\200\200\200\200.\200).
  Auto-renaming it.
  Renamed to FSCK0000.000
/apps/usr/lib/libpanel_g.a
  Bad short file name (\227\200\200\200\200\200\200.\200).
  Auto-renaming it.
  Renamed to FSCK0000.000
/apps/usr/lib/libform.so.5.9
  Bad short file name (\230\200\200\200\200\200\200.\200\203).
  Auto-renaming it.
  Renamed to FSCK0000.001
/apps/usr/lib/opkg/info/libiconv.control
  Bad short file name (\214\200\200\200\200\200\200.\200\203).
  Auto-renaming it.
  Renamed to FSCK0000.000
/apps/usr/include/ncurses.h
  Bad short file name (\237\200\200\200\200\200\200.\200).
  Auto-renaming it.
  Renamed to FSCK0000.000
/apps/usr/include/openssl/pem2.h
  Bad short file name (\236\200\200\200\200\200\200.\200\203).
  Auto-renaming it.
  Renamed to FSCK0000.000
/apps/usr/share/nano/patch.nanorc
  Bad short file name (\237\200\200\200\200\200\200.\200).
  Auto-renaming it.
  Renamed to FSCK0000.000
/apps/tmp/usr/lib/opkg/status
  File size is 0 bytes, cluster chain length is > 0 bytes.
  Truncating file to 0 bytes.
/apps/mnt/usr/lib/opkg/status
  File size is 0 bytes, cluster chain length is > 0 bytes.
  Truncating file to 0 bytes.
Checking for unused clusters.
Reclaimed 1 unused cluster (4096 bytes).
Checking free cluster summary.
Free cluster summary wrong (125446 vs. really 103709)
  Auto-correcting.
Leaving filesystem unchanged.
/dev/loop0p2: 487 files, 21737/125446 clusters

**Comment by ryniker** [comment]
A few characters were mangled in the fsck.vfat output posted above.  The original data may be
retrieved from  http://ryniker.org/minoca/fsck.vfat_log

**Comment by ryniker** [comment]
>Emacs does seem to rely on hard links doesn't it.

Not very much.  This code is from emacs-25.1/src/fileio.c in the
add-name-to-file function which creates another name (link) for an
existing file:

  unlink (SSDATA (newname));
  if (link (SSDATA (encoded_file), SSDATA (encoded_newname)) < 0)
    {
      int link_errno = errno;
      report_file_errno ("Adding new name", list2 (file, newname), link_errno);
    }

It is used by emacs-25.1/lisp/loadup.el in the last step of the emacs
build operation, to create a link from emacs-25.1.1 to emacs (the just
created executable file).  I do not think it wise to change the
semantics of add-name-to-file to fall back to symbolic links.  Better
just to create a symlink or copy the file when building emacs for
Minoca, and let other applications see the "Adding new name ..." error
and decide what is right for them to do.

**Comment by evangreen** [comment]
Hm, I wonder why fsck is mangling all our short names. According to this [wikipedia page](https://en.wikipedia.org/wiki/Design_of_the_FAT_file_system) (which I realize is not a canonical source) valid characters include 128-255 (with an exception for 0xe5 in the first position). I took a look at bad_name in dosfstools and didn't immediately see what's causing them to complain. 

The 'cluster chain length > 0' is expected behavior. We identify i-node numbers on FAT using cluster numbers. We need to assign clusters even to empty files so that we can assign consistent i-node numbers to those files. Slightly out of spec but no implementation seems to mishandle it.

The pagefile.sys cluster chain size, 1 unused cluster, and incorrect cluster summary could be bugs, or they might just be the result of an unclean shutdown. Unsure there.

Regarding emacs, I do recall some issues similar to what you mentioned. I think our diffs address something like that, not sure if it's exactly what you're describing.

I am still working on emacs, I created a separate issue for it in Github. I think I got the Control+Space issue sorted out. Need to understand the time crash you mentioned next.
